samza-api/src/main/java/org/apache/samza/operators/descriptors/base/stream/InputDescriptor.java

```diff
@@ -0,0 +1,218 @@
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.samza.operators.descriptors.base.system.ExpandingSystemDescriptor;
+import org.apache.samza.operators.descriptors.base.system.SystemDescriptor;
+import org.apache.samza.operators.descriptors.base.system.TransformingSystemDescriptor;
+import org.apache.samza.operators.functions.InputTransformer;
+import org.apache.samza.serializers.Serde;
+import org.apache.samza.system.SystemStreamMetadata.OffsetType;
+
+/**
+ * The base descriptor for an input stream. Allows setting properties that are common to all input streams.
+ *
+ * @param <StreamMessageType> type of messages in this stream.
+ * @param <SubClass> type of the concrete sub-class
+ */
+@SuppressWarnings("unchecked")
+public abstract class InputDescriptor<StreamMessageType, SubClass extends InputDescriptor<StreamMessageType, SubClass>>
+    extends StreamDescriptor<StreamMessageType, SubClass> {
+  private static final String RESET_OFFSET_CONFIG_KEY = "streams.%s.samza.reset.offset";
+  private static final String OFFSET_DEFAULT_CONFIG_KEY = "streams.%s.samza.offset.default";
+  private static final String PRIORITY_CONFIG_KEY = "streams.%s.samza.priority";
+  private static final String BOOTSTRAP_CONFIG_KEY = "streams.%s.samza.bootstrap";
+  private static final String BOUNDED_CONFIG_KEY = "streams.%s.samza.bounded";
+  private static final String DELETE_COMMITTED_MESSAGES_CONFIG_KEY = "streams.%s.samza.delete.committed.messages";
+
+  private Optional<InputTransformer<StreamMessageType>> transformerOptional = Optional.empty();
+  private Optional<Boolean> resetOffsetOptional = Optional.empty();
+  private Optional<OffsetType> offsetDefaultOptional = Optional.empty();
+  private Optional<Integer> priorityOptional = Optional.empty();
+  private Optional<Boolean> isBootstrapOptional = Optional.empty();
+  private Optional<Boolean> isBoundedOptional = Optional.empty();
+  private Optional<Boolean> shouldDeleteCommittedMessagesOptional = Optional.empty();
+
+  /**
+   * Constructs an {@link InputDescriptor} instance.
+   *
+   * @param streamId id of the stream
+   * @param systemName system name for the stream
+   * @param serde serde for messages in the stream
+   * @param systemDescriptor system descriptor this stream descriptor was obtained from if available, else null
+   * @param transformer stream level input stream transform function if available, else null
+   */
+  public InputDescriptor(String streamId, String systemName, Serde serde,
+      SystemDescriptor systemDescriptor, InputTransformer<StreamMessageType> transformer) {
+    super(streamId, systemName, serde, systemDescriptor);
+
+    // stream level transformer takes precedence over system level transformer
+    if (transformer != null) {
+      this.transformerOptional = Optional.of(transformer);
+    } else {
+      if (systemDescriptor instanceof TransformingSystemDescriptor) {
```

Comment:
If someone adds another type of descriptor which has "transforming" functionality, then they will need to remember to update this if-else check. You already got into this situation with ExpandingSystemDescriptor, which has a transformer, but it isn't obvious that it is also "transforming" unless you look at implementation.
My preference is to have a way to avoid changing this check in the future. The "optional transformer field" has the least code I think. Or, if you can find a way to make any descriptor with transforming functionlity extend TransformingSystemDescriptor, then you could still keep the transformation logic with TransformingSystemDescriptor and you will only ever need one special instanceof check for TransformingSystemDescriptor.